controllers/metal3.io/baremetalhost_controller.go

```diff
@@ -966,45 +966,50 @@ func (r *BareMetalHostReconciler) actionPreparing(prov provisioner.Provisioner,
 	// is invalid and needs to be reconfigured.
 	if info.host.Status.ErrorType == metal3v1alpha1.PreparationError {
 		prepareData.ActualRAIDConfig = nil
-		dirty = true
+		bmhDirty = true
 	}
 
 	// Use settings in hostFirmwareSettings if available
-	hfs, err := r.getHostFirmwareSettings(info)
+	hfsDirty, hfs, err := r.getHostFirmwareSettings(info)
+
 	if err != nil {
 		// wait until hostFirmwareSettings are ready
 		return actionContinue{subResourceNotReadyRetryDelay}
 	}
-	if hfs != nil {
+	if hfsDirty {
 		prepareData.ActualFirmwareSettings = hfs.Status.Settings.DeepCopy()
 		prepareData.TargetFirmwareSettings = hfs.Spec.Settings.DeepCopy()
 	}
 
-	provResult, started, err := prov.Prepare(prepareData, dirty)
+	provResult, started, err := prov.Prepare(prepareData, bmhDirty || hfsDirty, bmhDirty)
```

Comment:
I just realised I've been massively overthinking this.

The `force` flag now works exactly the same as all the other Provisioner APIs that have one. So we can pass the equivalent thing, which would be `info.host.Status.ErrorType == metal3v1alpha1.PreparationError`.

I think passing `bmhDirty` has a bug in the case where we are retrying but there is no difference between the Firmware/RAID settings in the BMH Spec and the empty ones in the Status.